### PR TITLE
Feature：Add prometheus to monitor the gateway and grpc server

### PR DESCRIPTION
**Pull request by Mutezebra**
Etcd-related methods in discover package are not reused in order to meet customization needs
Even it have a lot of change but most change is from json file，don't worry
See pkg/prometheus/README.md for more datail #68

### CODE REVIEW

File: app/gateway/rpc/init.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ import (
 	"log"
 	"time"
 
+	"github.com/CocaineCong/tangseng/pkg/prometheus"
+
```
Comment (CocaineCong) [inline app/gateway/rpc/init.go:27]: u can resort the sort of pkg

File: app/favorite/cmd/main.go [hunk 2]
```diff
@@ -60,10 +63,16 @@ func main() {
 		}
 	}()
 	handler := otelgrpc.NewServerHandler()
-	server := grpc.NewServer(grpc.StatsHandler(handler))
+	server := grpc.NewServer(
+		grpc.StatsHandler(handler),
+		grpc.UnaryInterceptor(prometheus.UnaryServerInterceptor),
+		grpc.StreamInterceptor(prometheus.StreamServerInterceptor),
```
Comment (CocaineCong) [inline app/favorite/cmd/main.go:69]: can u add some code comments?

File: consts/prometheus.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package consts
+
```
Comment (CocaineCong) [inline consts/prometheus.go:2]: look the file CONTRIBUTING_CN.md in root path before u make a pr

File: pkg/prometheus/etcd.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+type Instance struct {
+	Conf []*Conf
+}
+
+type Conf struct {
+	Targets []string          `json:"targets"`
+	Labels  map[string]string `json:"labels"`
+}
```
Comment (CocaineCong) [inline pkg/prometheus/etcd.go:21]: add some code comments for the mean of fields

File: pkg/prometheus/handler.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+)
+
+func GatewayHandler() gin.HandlerFunc {
+	EtcdRegister(config.Conf.Server.Metrics, consts.GatewayJobForPrometheus)
+	handler := promhttp.Handler()
+	return func(c *gin.Context) {
+		handler.ServeHTTP(c.Writer, c.Request)
+	}
+}
+
```
Comment (CocaineCong) [inline pkg/prometheus/handler.go:21]: add some code comments

File: pkg/prometheus/config/files/.gitkeep
[the remark this answers is not in the document]
Comment (Mutezebra) [reply pkg/prometheus/config/files/.gitkeep:1]: 因为git无法追踪文件夹，为了让这个文件夹被追踪以挂载到prometheus容器中，所以创建一个空文件。提前有这个文件夹存在的话也不必在代码中再去检查这个文件夹是否存在，以便于后续的json文件的生成。事实上prometheus是可以以静态文件的方式追踪grpc服务的metrics的，但是为了扩展性，我觉得以文件的动态发现的方式好一点。这样就不用手动的修改prometheus的配置文件从而实现服务的动态发现。

**Review by CocaineCong — APPROVED** [review]
LGTM

**Pull request merged**